rust/arrow/src/array/array.rs

```diff
@@ -56,11 +55,13 @@ pub trait Array: fmt::Debug + Send + Sync + JsonEqual {
     /// ```
     fn as_any(&self) -> &Any;
 
-    /// Returns a reference-counted pointer to the underlying data of this array.
-    fn data(&self) -> ArrayDataRef;
+    /// Returns a reference to the underlying data of this array.
+    fn data(&self) -> &ArrayData;
```

Comment:
if both `data()` and `data_ref()` return `&ArrayData` may be just `data_ref()` is enough and `data()` should be removed; or possibly change `data()` to return `ArrayData`

Reply:
I understand why we wanted to avoid `Array::data().clone()` to get `ArrayData`, but with these 2 functions returning the same reference, I'd opt to make `Array::data() -> ArrayData`

Reply:
@yordan-pavlov I think it's safer to deprecate `data()`, then remove it later. We still use it a lot in the codebase, but it's often more performant to use `array_ref()`, so returning `ArrayData` doesn't guide users on a faster path.

Reply:
yes, deprecate `data()` is what I meant, immediate removal is not great for backwards compatibility